### PR TITLE
Do not blow up on bad form input

### LINKED ISSUE #86: We should have some basic form validations
For example if somebody puts a non-parseable date in a field, we should probably re-render the page with an error at that field instead of blowing up.

### PR DESCRIPTION
**Pull request by singpolyma**
Show validation errors to user

Ideally these will be caught before sumission, but if they are not we shouldn't just blow up.  At least show something useful.

Closes #86

### CODE REVIEW

File: app/controllers/file.rb [hunk 1]
```diff
@@ -152,27 +152,31 @@ def slim(template, options={})
...
-						dir.join(resource.relative_path).open("w") do |f|
-							if updated_data
-								if resource.yml?
-									f.puts updated_data.to_yaml.sub(/\A---\r?\n?/, "")
-								else
-									f.puts updated_data.to_yaml
-									f.puts "---"
+					begin
+						site.commit_with("Modified #{resource.relative_path}", pundit_user) do |dir|
+								updated_data = FormHelpers.process(params[:data], dir)
```
Comment (ericroberts) [inline app/controllers/file.rb]: This should be outdented. Confused me for a second.

File: app/views/inputs/date.slim [hunk 2]
```diff
@@ -1 +1 @@
-input type="date" name=name value=(value && value.strftime("%F"))
+input type="date" name=name pattern="(?:19|20)[0-9]{2}-(?:(?:0[1-9]|1[0-2])-(?:0[1-9]|1[0-9]|2[0-9])|(?:(?!02)(?:0[1-9]|1[0-2])-(?:30))|(?:(?:0[13578]|1[02])-31))" value=(value && value.strftime("%F"))
```
Comment (ericroberts) [inline app/views/inputs/date.slim]: Should we move this pattern to a constant somewhere?

File: app/views/inputs/datetime.slim [hunk 3]
```diff
@@ -1,2 +1,2 @@
-input type="date" name="#{name}[___datetime_date]" value=(value && value.strftime("%F"))
-input type="time" name="#{name}[___datetime_time]" value=(value && value.strftime("%H:%M:%S"))
+input type="date" name="#{name}[___datetime_date]" pattern="(?:19|20)[0-9]{2}-(?:(?:0[1-9]|1[0-2])-(?:0[1-9]|1[0-9]|2[0-9])|(?:(?!02)(?:0[1-9]|1[0-2])-(?:30))|(?:(?:0[13578]|1[02])-31))" value=(value && value.strftime("%F"))
+input type="time" name="#{name}[___datetime_time]" pattern="(0?[0-9]|1[0-9]|2[0-3]):([0-9]|[0-5][0-9])(?::([0-9]|[0-5][0-9]))?" value=(value && value.strftime("%H:%M:%S"))
```
Comment (ericroberts) [inline app/views/inputs/datetime.slim]: Should we move this pattern to a constant somewhere?

File: app/form_helpers.rb [hunk 4]
```diff
@@ -68,6 +70,8 @@ def checkbox
 			def datetime
 				datetime = "#{self["___datetime_date"]} #{self["___datetime_time"]}"
 				Time.parse(datetime) if datetime.strip.to_s != ""
+			rescue ArgumentError
```
Comment (ericroberts) [inline app/form_helpers.rb]: A test to assert that this happens might be good?

File: app/form_helpers.rb [hunk 5]
```diff
@@ -76,7 +80,12 @@ def array
 
 			def convert_dates
 				merge(self) do |k,v|
-					v = Date.parse(v) if is_date?(k, v)
+					begin
+						v = Date.parse(v) if is_date?(k, v)
+					rescue ArgumentError
```
Comment (ericroberts) [inline app/form_helpers.rb]: May want to test that this happens.

**Comment by singpolyma** [comment]
@ericroberts tests and constants added

**Pull request merged**